src/decoders/citrix_ctx1_decoder.rs

```diff
@@ -0,0 +1,187 @@
+use crate::checkers::CheckerTypes;
+use crate::decoders::interface::check_string_success;
+
+use super::crack_results::CrackResult;
+use super::interface::Crack;
+use super::interface::Decoder;
+
+use log::{debug, info, trace};
+
+///! Citrix CTX1 Decoder
+pub struct CitrixCTX1Decoder;
+
+impl Crack for Decoder<CitrixCTX1Decoder> {
+    fn new() -> Decoder<CitrixCTX1Decoder> {
+        Decoder {
+            name: "citrix_ctx1",
+            description: "Citrix CTX1 is a very old encoding that was used for encoding Citrix passwords.",
+            link: "https://www.remkoweijnen.nl/blog/2012/05/13/encoding-and-decoding-citrix-passwords/",
+            tags: vec!["citrix_ctx1", "citrix", "passwords", "decoder"],
+            expected_runtime: 0.01,
+            expected_success: 1.0,
+            failure_runtime: 0.01,
+            normalised_entropy: vec![1.0, 10.0],
+            popularity: 0.1,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// This function does the actual decoding
+    /// It returns an Option<string> if it was successful
+    /// Else the Option returns nothing and the error is logged in Trace
+    fn crack(&self, text: &str, checker: &CheckerTypes) -> CrackResult {
+        trace!("Trying citrix_ctx1 with text {:?}", text);
+        let decoded_text: Option<String> = decode_citrix_ctx1(text);
+
+        trace!("Decoded text for citrix_ctx1: {:?}", decoded_text);
+        let mut results = CrackResult::new(self, text.to_string());
+
+        if decoded_text.is_none() {
+            debug!("Failed to decode citrix_ctx1 because the length is not a multiple of 4");
+            return results;
+        }
+
+        let decoded_text = decoded_text.unwrap();
+        if !check_string_success(&decoded_text, text) {
+            info!(
+                "Failed to decode citrix_ctx1 because check_string_success returned false on string {}",
+                decoded_text
+            );
+            return results;
+        }
+
+        let checker_result = checker.check(&decoded_text);
+        results.unencrypted_text = Some(decoded_text);
+
+        results.update_checker(&checker_result);
+
+        results
+    }
+}
+
+/// Decodes Citrix CTX1
+fn decode_citrix_ctx1(text: &str) -> Option<String> {
+    if text.len() % 4 != 0 {
+        return None;
+    }
+
+    if text.to_uppercase() != text || !text.chars().all(|c| c.is_ascii()) {
+        return None;
+    }
+
+    let mut rev = text.as_bytes().to_vec();
+    rev.reverse();
+    let mut result = Vec::new();
+    let mut temp;
```

Comment:
Bet you could reduce these nuts a bit 👀

Reply:
muts

Reply:
Yes, this decoder can be cleaned up. I'll do it in the future.